src/main/java/uk/gov/digital/ho/hocs/entrypoint/CsvHelpers.java

```diff
@@ -0,0 +1,78 @@
+package uk.gov.digital.ho.hocs.entrypoint;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+@Slf4j
+public class CsvHelpers {
+    public static Stream<Map<String, String>> getRowsFromInputStream(InputStream inputStream) throws IOException {
+        CsvSchema bootstrapSchema = CsvSchema.emptySchema().withHeader();
+        CsvMapper mapper = new CsvMapper();
+
+        MappingIterator<Map<String, String>> readValues =
+            mapper.readerFor(Map.class).with(bootstrapSchema).readValues(inputStream);
+
+        return StreamSupport.stream(
+            Spliterators.spliteratorUnknownSize(readValues, 0),
+            false
+        );
+    }
+
+    public static StreamingResponseBody streamCsvToResponseBody(
+        Stream<Map<String, String>> rows,
+        List<String> columnOrdering
+    ) {
+        return outputStream -> {
+            //noinspection ResultOfMethodCallIgnored
+            rows.reduce(
+                (JsonGenerator) null,
+                (generator, row) -> {
+                    try {
+                        if (generator == null) {
```

Comment:
Could generator be created outside the reduce so you don't have to build it on the first iteration?

Reply:
Because the row contains arbitrary case data fields, the full column list isn't known until an item in the stream is available, and there isn't a way to get that item first without consuming the stream. This was the least awkward way I could get it to work, but if you know of a better pattern I'd be happy to change it.